src/core/__tests__/ReactCompositeComponent-test.js

```diff
@@ -327,10 +328,39 @@ describe('ReactCompositeComponent', function() {
       instance.setState({ value: 2 });
     }).toThrow(
       'Invariant Violation: setState(...): Can only update a mounted or ' +
-      'mounting component.'
+      'mounting component. This usually means you called setState() on an ' +
+      'unmounted component.'
     );
   });
 
+  it('should not allow `setState` on unmounting components', function() {
+    var container = document.createElement('div');
+    document.documentElement.appendChild(container);
+
+    var Component = React.createClass({
+      getInitialState: function() {
+        return { value: 0 };
+      },
+      componentWillUnmount: function() {
+        expect(() => this.setState({ value: 2 })).toThrow(
+          'Invariant Violation: replaceState(...): Cannot update while ' +
+          'unmounting component. This usually means you called setState() ' +
+          'on an unmounted component.'
+        );
+      },
+      render: function() {
+        return <div />;
+      }
+    });
+
+    var instance = React.render(<Component />, container);
+    expect(function() {
+      instance.setState({ value: 1 });
+    }).not.toThrow();
+
+    React.unmountComponentAtNode(container);
```

Comment:
We don't actually unmount in most of our tests. I think you could skip this, as well as appending container to the DOM (non of our tests do this, just having the node is good enough)